lib/engine/game/g_1862/game.rb

```diff
@@ -1512,9 +1523,502 @@ def raise_money!(entity, amount)
+            @merge_data[:price] = new_price
+            @merge_data[:par] = new_par_price
+            old_price.corporations.delete(survivor)
+            new_price.corporations << survivor
+            survivor.share_price = new_price
+            survivor.par_price = new_par_price
+            survivor.original_par_price = new_par_price
+          end
+
+          # 2. move any chartered IPO shares to market
+          corps.each do |corp|
+            if @chartered[corp] && !(ipo_shares = corp.ipo_shares).empty?
+              ipo_shares.each { |s| transfer_share(s, @share_pool) }
+              @log << "Moved #{ipo_shares.size} shares from #{corp.name} IPO to market"
+            end
+          end
+
+          # 3. move all treasury shares to originator
+          corps.each do |corp|
+            next if corp == originator
+
+            corp.shares_of(corp).dup.each { |s| transfer_share(s, originator) }
+          end
+
+          # 4. return half of shares, possibly offer to redeem or sell options (half-shares)
+          return if @merge_data[:holders].any? { |holder| return_half_and_swap(holder) }
+
+          merge_post_shares # no need to ask player
+        end
+
+        # called by step after player makes choice about option
+        def continue_merge_option(entity)
+          remaining_holders = @merge_data[:holders][(@merge_data[:holders].index(entity) + 1)..-1]
+          return if remaining_holders.any? { |holder| return_half_and_swap(holder) }
+
+          merge_post_shares # no need to ask player
+        end
+
+        # always return survivor shares first and director's first within those
+        def affected_shares(entity, corps)
+          affected = entity.shares.select { |s| s.corporation == corps.first }.sort_by(&:percent).reverse
+          unless corps.one?
+            affected.concat(entity.shares.select { |s| s.corporation == corps.last }.sort_by(&:percent).reverse)
+          end
+          affected
+        end
+
+        # reorder shares
+        # 1. nonsurviving director cert
+        # 2. nonsurviving plain certs
+        # 3. surviving plain certs
+        # 4. surviving director cert
+        def reorder_shares(shares, corps)
+          if corps.one?
+            shares.select { |s| s.corporation == corps.first }.sort_by(&:percent)
+          else
+            shares.select { |s| s.corporation == corps.last }.sort_by(&:percent).reverse +
+              shares.select { |s| s.corporation == corps.first }.sort_by(&:percent)
+          end
+        end
+
+        # look for next entity that has needed share
+        def find_donor_share(entity, corp, holders, percent)
+          return nil if entity.corporation?
+
+          donors = [@share_pool] + holders[(holders.find_index(entity) + 1)..-1]
+
+          donors.each do |holder|
+            match = holder.shares_of(corp).find { |s| s.percent == percent }
+            return match if match
+          end
+          nil
```

Comment:
Maybe it works? https://opalrb.com/docs/api/v0.7.2/corelib/Enumerator/Lazy.html

Reply:
@roseundy test this out if you can before making this change

Reply:
This hurts my brain. I need to look this up.

Reply:
* map everything to what you called `match`
* `itself` is a no-op, `find` returns the first truthy element, so `.find(&:itself)` returns the first truthy element in the Array

I don't know the inner implementation details, but `lazy` somehow executes the whole chain of methods for each element, and will stop early if the last method would stop, rather than executing each method against the whole list:

```
irb(main):052:0> x = [2,3,4,5]
irb(main):053:0> x.lazy.map { |a| puts "a = #{a}"; a % 3 == 0 }.find(&:itself)
a = 2
a = 3
=> true
```

Reply:
I'll give it a shot.

Reply:
Nope, it didn't work.

Reply:
Works w/o the .lazy, but not what we want.

Reply:
Yeah just leave it as is then. 👍